pkg/prober/prober.go

```diff
@@ -0,0 +1,157 @@
+	router.Get(healthyEndpointPath, p.probeHandlerFunc(p.IsHealthy, "healthy"))
+	router.Get(readyEndpointPath, p.probeHandlerFunc(p.IsReady, "ready"))
+}
+
+// RegisterInMux registers readiness and liveness probes to mux.
+func (p *Prober) RegisterInMux(mux *http.ServeMux) {
+	mux.HandleFunc(healthyEndpointPath, p.probeHandlerFunc(p.IsHealthy, "healthy"))
+	mux.HandleFunc(readyEndpointPath, p.probeHandlerFunc(p.IsReady, "ready"))
+}
+
+func (p *Prober) writeResponse(w http.ResponseWriter, probeFunc func() error, probeType string) {
+	err := probeFunc()
+	if err != nil {
+		http.Error(w, fmt.Sprintf(errorProbeText, p.getComponent(), probeType, err), probeErrorHTTPStatus)
+		return
+	}
+	if _, e := io.WriteString(w, fmt.Sprintf(okProbeText, p.getComponent(), probeType)); e == nil {
+		level.Error(p.getLogger()).Log("msg", "failed to write probe response", "probe type", probeType, "err", err)
+	}
+}
+
+func (p *Prober) probeHandlerFunc(probeFunc func() error, probeType string) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, _ *http.Request) {
+		p.writeResponse(w, probeFunc, probeType)
+	}
+}
+
+// IsReady returns error if component is not ready and nil if it is.
+func (p *Prober) IsReady() error {
+	p.readyMtx.Lock()
+	defer p.readyMtx.Unlock()
+	return p.readiness
+}
+
+// SetReady sets components status to ready.
+func (p *Prober) SetReady() {
+	if p.IsReady() != nil {
+		level.Info(p.getLogger()).Log("msg", "changing probe status", "status", "ready")
+	}
+	p.SetNotReady(nil)
+}
+
+// SetNotReady sets components status to not ready with given error as a cause.
+func (p *Prober) SetNotReady(err error) {
+	if err != nil && p.IsReady() == nil {
+		level.Warn(p.getLogger()).Log("msg", "changing probe status", "status", "not-ready", "reason", err)
+	}
+	p.readyMtx.Lock()
+	defer p.readyMtx.Unlock()
+	p.readiness = err
+}
+
+// IsHealthy returns error if component is not healthy and nil if it is.
+func (p *Prober) IsHealthy() error {
+	p.healthyMtx.Lock()
+	defer p.healthyMtx.Unlock()
+	return p.healthiness
+}
+
+// SetHealthy sets components status to healthy.
+func (p *Prober) SetHealthy() {
+	if p.IsHealthy() != nil {
+		level.Info(p.getLogger()).Log("msg", "changing probe status", "status", "healthy")
+	}
+	p.SetNotHealthy(nil)
+}
+
+// SetNotHealthy sets components status to not healthy with given error as a cause.
+func (p *Prober) SetNotHealthy(err error) {
+	if err != nil && p.IsHealthy() == nil {
+		level.Warn(p.getLogger()).Log("msg", "changing probe status", "status", "unhealthy", "reason", err)
+	}
+	p.healthyMtx.Lock()
+	defer p.healthyMtx.Unlock()
+	p.healthiness = err
+}
+
+// HandleIfReady if probe is ready calls the function otherwise returns 503.
+func (p *Prober) HandleIfReady(f http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		ready_err := p.IsReady()
+		if ready_err == nil {
+			f(w, r)
+			return
+		}
+		p.writeResponse(w, p.IsReady, "ready")
```

Comment:
There's a small error here. By the time you call this. `p.IsReady()` might start indicating that it is suddenly ready, right? AFAICT you need to do both of these actions while `p.readyMtx` is locked.

Reply:
Good point! The chances are really small but still this is a race. Thanks!

Reply:
PTAL if this way it's ok with you